coderd/workspaceagentsrpc_internal_test.go

```diff
@@ -0,0 +1,376 @@
+		conn:              fConn,
+		db:                mDB,
+		replicaID:         replicaID,
+		updater:           fUpdater,
+		logger:            logger,
+		pingPeriod:        testutil.IntervalFast,
+		disconnectTimeout: testutil.WaitShort,
+	}
+	uut.init()
+
+	connected := mDB.EXPECT().UpdateWorkspaceAgentConnectionByID(
+		gomock.Any(),
+		connectionUpdate(agent.ID, replicaID),
+	).
+		AnyTimes().
+		Return(nil)
+	mDB.EXPECT().UpdateWorkspaceAgentConnectionByID(
+		gomock.Any(),
+		connectionUpdate(agent.ID, replicaID, withDisconnected()),
+	).
+		After(connected).
+		Times(1).
+		Return(nil)
+	mDB.EXPECT().GetLatestWorkspaceBuildByWorkspaceID(gomock.Any(), build.WorkspaceID).
+		AnyTimes().
+		Return(database.WorkspaceBuild{ID: build.ID}, nil)
+
+	closeCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	done := make(chan struct{})
+	go func() {
+		uut.mind(closeCtx)
+		close(done)
+	}()
+	// wait a couple intervals, but not long enough for a disconnect
+	time.Sleep(3 * testutil.IntervalFast)
+	fConn.requireNotClosed(t)
+	fUpdater.requireEventuallySomeUpdates(t, build.WorkspaceID)
+	n := fUpdater.getUpdates()
+	cancel()
+	fConn.requireEventuallyClosed(t, websocket.StatusGoingAway, "canceled")
+
+	// make sure we got at least one additional update on close
+	_ = testutil.RequireRecvCtx(ctx, t, done)
+	m := fUpdater.getUpdates()
+	require.Greater(t, m, n)
+}
+
+func TestAgentWebsocketMinder_PingTimeout(t *testing.T) {
+	t.Parallel()
+	ctx := testutil.Context(t, testutil.WaitShort)
+	now := dbtime.Now()
+	fConn := &fakePingerCloser{}
+	ctrl := gomock.NewController(t)
+	mDB := dbmock.NewMockStore(ctrl)
+	fUpdater := &fakeUpdater{}
+	logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+	agent := database.WorkspaceAgent{
+		ID: uuid.New(),
+		FirstConnectedAt: sql.NullTime{
+			Time:  now.Add(-time.Minute),
+			Valid: true,
+		},
+	}
+	build := database.WorkspaceBuild{
+		ID:          uuid.New(),
+		WorkspaceID: uuid.New(),
+	}
+	replicaID := uuid.New()
+
+	uut := &agentWebsocketMinder{
+		apiCtx:            ctx,
+		workspaceAgent:    agent,
+		workspaceBuild:    build,
+		conn:              fConn,
+		db:                mDB,
+		replicaID:         replicaID,
+		updater:           fUpdater,
+		logger:            logger,
+		pingPeriod:        testutil.IntervalFast,
+		disconnectTimeout: testutil.WaitShort,
+	}
+	uut.init()
+	// set the last ping to the past, so we go thru the timeout
+	uut.lastPing.Store(ptr.Ref(now.Add(-time.Hour)))
+
+	connected := mDB.EXPECT().UpdateWorkspaceAgentConnectionByID(
+		gomock.Any(),
+		connectionUpdate(agent.ID, replicaID),
+	).
+		AnyTimes().
+		Return(nil)
+	mDB.EXPECT().UpdateWorkspaceAgentConnectionByID(
+		gomock.Any(),
+		connectionUpdate(agent.ID, replicaID, withDisconnected()),
+	).
+		After(connected).
+		Times(1).
+		Return(nil)
+	mDB.EXPECT().GetLatestWorkspaceBuildByWorkspaceID(gomock.Any(), build.WorkspaceID).
+		AnyTimes().
+		Return(database.WorkspaceBuild{ID: build.ID}, nil)
+
+	go uut.mind(ctx)
```

Comment:
Relying on context to shut down this method its racy in that we may trigger goleak sporadically. Might be best to synchronize this to the test.

Reply:
goleak waits for goroutines to complete up to a timeout. I don't believe we are in danger of triggering it unless the routine deadlocks, in which case goleak will have found a legit bug.

Reply:
By default for only 2 seconds I believe, and I don’t think it’s configurable. Plenty of time for a slow env like Windows runner to trigger the edge case, I’d say.

Reply:
Even on a slow Windows runner, 2 seconds is absolutely ages.  I've never seen goleak give a  false positive on a goroutine that was unblocked and just waiting to be scheduled by the runtime and to finish its work. Have you?

Reply:
I have seen such failures in the past, which is why I raised the concern. In my experience it can and has resulted in rare flakes. I'd like to concretely say which scenarios, but truth be told I don't remember.

I also can't say whether or not it's the case for the current Windows runners, the only thing I can clearly say is that for the old GH Actions Windows runner 2 seconds was definitely not a long time and such expectations often resulted in a flake. 😄 We run so many things in parallel it's unclear how much time is needed, the more things exit leaving other things behind, the closer we get to filling up that 2 second gap, I suppose.